tests/integration/utils.py

```diff
@@ -463,3 +465,167 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
         dep_file = output_dir / "deps" / path
         assert dep_file.exists()
         assert dep_file.read_text() == expected_content
+
+
+ALL = object()
+# Individual files could be added to the set as well.
+paths_to_code = set((Path("cachi2"), Path("tests/integration")))
+supported_pms = set(list(resolver._package_managers) + list(resolver._dev_package_managers))
+
+
+def retrieve_changed_files_from_git(ref: str = "HEAD") -> tuple[Path, ...]:
+    repo = Repo(".", search_parent_directories=True)
+    commit = repo.commit(ref)
+    modified_files = [Path(f) for f in commit.stats.files]
+    return tuple(modified_files)
```

Comment:
I noticed that the function returns changes only made in the last commit.
I suppose it should return changes from `<current_branch>..main` or ?

Reply:
My intent was to run this for the latest commit only. Diffing feature branch with main would produce too many changed modules and would effectively negate all the benefits of the change. Suppose one is working on pip in one commit, then on gomod, then on npm. My change addresses the fact that it is not necessary to retest pip when gomod is affected.

Reply:
> This feels a bit like overengineering to me. It is extremely easy to trigger all tests to run

But that's exactly not what we want here. To be precise, the scenario where someone would touch all pkg mgrs at the same time is extremely rare. And I specifically don't want to run tests for pkg mgrs I'm not touching and I know that the related tests take too long to execute (gomod, pip), so if I'm touching both RPM and yarn in a single PR, then I don't want the tests to run just for yarn or RPM depending on what the last commit is. And I most certainly don't want to run tests for pip and gomod when I know they take twice as long.

> If this is the desired behavior then we already have an override variable for that. 

What variable do you mean?

I think the whole point of this PR was to kind of automate the detection of which tests need to be run for a particular change that would work both for one-shot executions as well in interactive rebase scenarios. At least that was my understanding and the value I was seeing in that, i.e. not needing to use any variable (unless absolutely needed for some heavy tweaking) and it'd just do the "right thing".

Reply:
What if the last commit in the PR updates only the README.md, but the previous commits are related to one or any number of package managers? We don't want to skip all integration tests in that case.

Reply:
The fact that GH tests all commits in a sequence (see a screenshot of the current sequence below)
![image](https://github.com/user-attachments/assets/eb7f0e82-f7db-4737-bf33-f5d5e9db76e2)
is proven by the latest (at the moment of writing) [IT job].(https://github.com/containerbuildsystem/cachi2/actions/runs/10781837227/job/29900605591?pr=612)

There you can observe that all tests were run (however second commit in the sequence does not affect any code) and there was no failure due to insufficient permissions despite the fact that the necessary permissions are added in the second commit, not the first one. Further, the job does this to fetch the code:
```
/usr/bin/git -c protocol.version=2 fetch --prune --no-recurse-submodules origin +refs/heads/*:refs/remotes/origin/* +refs/tags/*:refs/tags/* +aa22d80cf78f8d981813bfda262e8ba7f2319589:refs/remotes/pull/612/merge
```

and checks out the pull request like this:
```
 /usr/bin/git checkout --progress --force refs/remotes/pull/612/merge
```
This results in HEAD pointing to a merge commit:
```HEAD is now at aa22d80 Merge 7a546e32f90ad33e8e10d95e3164b777af957867 into b3862f2c08d62a901840f33a6a05eabaf46c0365
```
`7a546e3`  is my latest commit, `b3862f2` is the tip of `main` at the moment of the last push. It results in all files modified in the feature branch to be counted together.

Reply:
> Further, the job does this to fetch the code:
>```
>/usr/bin/git -c protocol.version=2 fetch --prune --no-recurse-submodules origin +refs/heads/*:refs/remotes/>origin/* +refs/tags/*:refs/tags/* +aa22d80cf78f8d981813bfda262e8ba7f2319589:refs/remotes/pull/612/merge
>```
>and checks out the pull request like this:
 >```
>/usr/bin/git checkout --progress --force refs/remotes/pull/612/merge
>```
>This results in HEAD pointing to a merge commit:
>```
>HEAD is now at aa22d80 Merge 7a546e32f90ad33e8e10d95e3164b777af957867 into b3862f2c08d62a901840f33a6a05eabaf46c0365
>```

I never noticed ^this, TIL GitHub automatically creates a "virtual" merge commit as a preview of the changes what it'd look like if the PR were to be merged to main immediately, thank you!

That said, what you modeled according to GitHub's behaviour won't be the case locally where your branch is properly rebased onto main and you won't have this automatic merge commit which you could inspect for changes. Instead, you'll have your individual commits and so based on your current implementation I am either: 
- forced to always run the integration test suite in an interactive rebase (which should be a choice really and a good practice, not a must) OR 
- set an environment variable to always run **ALL** tests because the last commit would end up yielding no relevant changes

none of which is desirable really because it's both not that much different from the status quo and arguably actually a regression in the behaviour because this solution only really tackles the gating environment but neglects local dev environments. If you instead compare all commits against main instead of just HEAD, then it should work consistently in both local and gating env.